### PR TITLE
implement update handler test cases

### PR DESCRIPTION
**Pull request by kmrmt**
Signed-off-by: Kosuke Morimoto <ksk@vdaas.org>

### Description:

SSIA

### Environment:

- Go Version: 1.18.2
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.14.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [X] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [X] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [X] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 1]
```diff
@@ -4005,6 +4005,50 @@ func Test_server_Update(t *testing.T) {
 		}
 		return nil
 	}
+
+	/*
+		Update test cases (only test float32 unless otherwise specified):
+		- Equivalence Class Testing ( 1000 vectors inserted before a update )
+			- case 1.1: success update vector
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4012]: 1 vector or all vectors?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 2]
```diff
@@ -4005,6 +4005,50 @@ func Test_server_Update(t *testing.T) {
...
+			- case 3.1: fail update with different dimension vector (type: uint8)
+			- case 3.2: fail update with different dimension vector (type: float32)
+		- Boundary Value Testing ( 1000 vectors inserted before a update )
+			- case 1.1: fail update with ""
+			- case 2.1: success update with ^@
+			- case 2.2: success update with ^I
+			- case 2.3: success update with ^J
+			- case 2.4: success update with ^M
+			- case 2.5: success update with ^[
+			- case 2.6: success update with ^?
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4023]: the value is set as Id?

Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4023]: the value is set as Id?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 1 above)
Comment (kmrmt) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4012]: It means 1 vector. I will make it explicit.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 2 above)
Comment (kmrmt) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4023]: Yes. These values are ID. I will make it explicit too.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 1 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: I think it is single update function to update 1 vector, we do not need to specify the number of the vector to be update.
What do you think?

Comment (kmrmt) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: hmm...
I think update is an unary operation, so I agree with you that it is not necessary to describe the number of vectors.
However, the meaning does not change whether to specify or not, I think there is no problem to do it.

Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: > I think it is single update function to update 1 vector, we do not need to specify the number of the vector to be update.
But, 1000 vectors inserted before update test, right :sweat_smile: ?

Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: @kevindiu :arrow_up:

Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: Yes, I mean it is a update function to update only 1 vector, it is redundant to specific the number of vector to be update IMO.
both ok for me.

**Comment by kmrmt** [comment]
@vankichi @kevindiu @kpango I have two questions that are below.
```
--- FAIL: Test_server_Update (0.03s)
    --- FAIL: Test_server_Update/Decision_Table_Testing_case_2.3:_fail_update_with_one_duplicated_vector,_different_ID_and_SkipStrictExistCheck_is_false (0.41s)
        handler_test.go:4880: error = got uuid: ""1"",
            				want uuid: """"
    --- FAIL: Test_server_Update/Boundary_Value_Testing_case_8.1:_success_update_with_one_NaN_value_vector_(type:_float32) (0.37s)
        handler_test.go:4880: error = got uuid: ""test"",
            				want uuid: """"
```

About the first case, I think that Decision Table Testing case 2.3 should fail but return success response.
About the second case, I think that Boundary Value Testing case 8.1 should  fail but return success response.
What do you think about them?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 3]
```diff
@@ -3972,90 +3972,900 @@ func Test_server_MultiInsert(t *testing.T) {
...
+				reqs[i] = &payload.Insert_Request{
+					Vector: &payload.Object_Vector{
+						Id:     strconv.Itoa(i),
+						Vector: v,
+					},
+					Config: &payload.Insert_Config{
+						SkipStrictExistCheck: true,
+					},
+				}
+			}
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4082]: I suppose we can use `genMultiInsertReq` function in https://github.com/vdaas/vald/pull/1612
could you please add `TODO` comment?
```suggestion
                        // TODO: use genMultiInsertReq()
			reqs := make([]*payload.Insert_Request, insertNum)
			for i, v := range gen(insertNum, cfg.Dimension) {
				reqs[i] = &payload.Insert_Request{
					Vector: &payload.Object_Vector{
						Id:     strconv.Itoa(i),
						Vector: v,
					},
					Config: &payload.Insert_Config{
						SkipStrictExistCheck: true,
					},
				}
			}
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 4]
```diff
@@ -3972,90 +3972,900 @@ func Test_server_MultiInsert(t *testing.T) {
...
+		}
+		return
+	}
+	fillUint8 := func(val uint8, size int) (ret []uint8) {
+		ret = make([]uint8, size)
+		for i := range ret {
+			ret[i] = val
+		}
+		return
+	}
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4120]: How about defining these as global functions in this test file?

Comment (kmrmt) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4120]: I agree with you.
In my opinion, these functions will be usable as utility functions for other handlers.
What do you think about implementing test utility package like pkg/testing?

Comment (kmrmt) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4120]: Sorry, I rethink that they implement to file global in this PR.
And if we have time, we will refactor them.

**Review by vankichi — APPROVED** [review]

**Review by kevindiu — APPROVED** [review]
